independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/builtin/beans/InjectionPointBuiltInBeanTest.java

```diff
@@ -0,0 +1,76 @@
+package io.quarkus.arc.test.builtin.beans;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Field;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Instance;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.spi.AnnotatedField;
+import javax.enterprise.inject.spi.AnnotatedParameter;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class InjectionPointBuiltInBeanTest {
+    @RegisterExtension
+    public ArcTestContainer container = ArcTestContainer.builder()
+            .beanClasses(MyProducer.class, MyService.class)
+            .additionalClasses(MyPojo.class)
+            .build();
+
+    @Test
+    public void test() {
+        MyService bean = Arc.container().select(MyService.class).get();
+        assertEquals("Hello MyService.pojo|MyProducer.produce(1)", bean.hello());
+    }
+
+    // ---
+
+    @Singleton
+    static class MyProducer {
+        @Produces
+        @Dependent
+        public MyPojo produce(InjectionPoint injectionPoint, Instance<Object> lookup) {
+            Field field = ((AnnotatedField<?>) injectionPoint.getAnnotated()).getJavaMember();
+            String f = field.getDeclaringClass().getSimpleName() + "." + field.getName();
+
+            // producer method parameters are injection points, so looking up `InjectionPoint` from `lookup`
+            // must return the injection point corresponding to the `lookup` producer method parameter
+            InjectionPoint lookupInjectionPoint = lookup.select(InjectionPoint.class).get();
```

Comment:
Hm, so you would get `null` for a non-`@Dependent` producer, right? 

I think that the programmatic lookup for `InjectionPoint` is undefined in the spec but it could be helpful in case of a `@Dependent` synthetic bean...

Reply:
Actually I'm wrong! This piece of code `lookup.select(InjectionPoint.class)` actually looks up an injection point of the `Instance` object.

For non-`@Dependent` producers, you should either get `null` in the `InjectionPoint` parameter of the producer, or perhaps that producer would be downright rejected -- not totally sure at the moment.

Reply:
> For non-@Dependent producers, you should either get null in the InjectionPoint parameter of the producer, or perhaps that producer would be downright rejected -- not totally sure at the moment.

I have a deja vu feeling we discussed this sometime ago :D
I can check tomorrow how it works for Weld if that helps but it might be outside of TCKs still

Reply:
> And creator functions of synthetic beans are exactly the motivation for this -- sorry for not making it clear. (More specifically, creator functions per the CDI Lite build compatible extensions. I'm not sure if ArC's synthetic bean creation API allows doing something like this.)

Do you mean https://github.com/eclipse-ee4j/cdi/blob/master/api/src/main/java/jakarta/enterprise/inject/build/compatible/spi/SyntheticBeanCreator.java#L44. ArC has https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/BeanCreator.java and I suppose it was a source of inspiration for BCE ;-).

> Actually I'm wrong! This piece of code lookup.select(InjectionPoint.class) actually looks up an injection point for the Instance object.

Hm, so what would you get if you call `lookup.select(InjectionPoint.class)` inside a `SyntheticBeanCreator#create()`?

> I can check tomorrow how it works for Weld if that helps but it might be outside of TCKs still

The behavior is IMO undefined and non-portable.

Reply:
> > And creator functions of synthetic beans are exactly the motivation for this -- sorry for not making it clear. (More specifically, creator functions per the CDI Lite build compatible extensions. I'm not sure if ArC's synthetic bean creation API allows doing something like this.)
> 
> Do you mean https://github.com/eclipse-ee4j/cdi/blob/master/api/src/main/java/jakarta/enterprise/inject/build/compatible/spi/SyntheticBeanCreator.java#L44. ArC has https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/BeanCreator.java and I suppose it was a source of inspiration for BCE ;-).

That's indeed what I mean, and yes, the ArC API was very much an inspiration for the BCE API (especially the parameters part). Except we turned it into something much closer to Portable Extensions :-)

> > Actually I'm wrong! This piece of code lookup.select(InjectionPoint.class) actually looks up an injection point for the Instance object.
> 
> Hm, so what would you get if you call `lookup.select(InjectionPoint.class)` inside a `SyntheticBeanCreator#create()`?

So there's a huge difference there. A parameter of a producer method is an injection point. If you inject `Instance` there and then lookup `InjectionPoint` from it, you'll get the injection point into which the `Instance` was injected. I believe you'll see the same behavior if you `@Inject Instance` anywhere else.

On the other hand, a parameter of the synthetic bean creation function is _not_ an injection point. That's why we provide `Instance<Object>` -- so you can simulate injection into the creation function by programmatic lookup. If you call `lookup.select(InjectionPoint.class)` inside a `SyntheticBeanCreator#create()`, you should get an injection point to which the currently created instance of a `@Dependent` synthetic bean is being injected. (If the synthetic bean is not `@Dependent`, then it's undefined.)

> > I can check tomorrow how it works for Weld if that helps but it might be outside of TCKs still
> 
> The behavior is IMO undefined and non-portable.

I believe the crux of this follows directly from the specification. Notably, https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#injection_point says:

> The container must provide a bean with scope `@Dependent`, bean type `InjectionPoint` and qualifier `@Default`, allowing dependent objects, as defined in [Dependent objects](https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#dependent_objects), to obtain information about the injection point to which they belong.

This is what this PR does. Other rules for injection and programmatic lookup cover the case of `Instance.select(InjectionPoint.class)`.

> If a bean that declares any scope other than `@Dependent` has an injection point of type `InjectionPoint` and qualifier `@Default`, the container automatically detects the problem and treats it as a definition error.

This makes it pretty clear that injecting an `InjectionPoint` into a non-`@Dependent` producer is invalid. There's indeed nothing in the spec (to the best of my knowledge) that covers what exactly should happen when looking up `InjectionPoint` when instantiating a non-`@Dependent` synthetic bean bean, but it IMHO follows from the above that it is invalid as well. See also https://github.com/eclipse-ee4j/cdi-tck/blob/master/impl/src/main/java/org/jboss/cdi/tck/tests/build/compatible/extensions/syntheticBeanInjectionPoint/SyntheticBeanInjectionPointTest.java#L32-L39

Reply:
> There's indeed nothing in the spec (to the best of my knowledge) that covers what exactly should happen when looking up `InjectionPoint` when instantiating a non-`@Dependent` synthetic bean bean, but it IMHO follows from the above that it is invalid as well.

And _invalid_ means exactly what? I mean an _invalid_ injection point is a definition error. But for lookup we would have to throw some exception or something like that...

Reply:
That's where I'll repeat that, while I believe it follows from the spec that it's invalid, the spec doesn't really say what should happen. In my opinion, returning `null` or throwing an exception are both acceptable :-) In my ArC fork where I implement BCE, I throw an exception, and if I remember correctly, Weld does the same.

But that's for a future PR, if you ask me :-)